File: ide-features-trainer/src/training/commands/MoveCaretCommand.java
package training.commands;

import com.intellij.find.FindManager;
import com.intellij.find.FindModel;
import com.intellij.find.FindResult;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.LogicalPosition;
import com.intellij.openapi.project.Project;
import org.jdom.Element;
import java.util.concurrent.ExecutionException;

/**
 * Created by karashevich on 30/01/15.
 */
public class MoveCaretCommand extends Command {

    final public static String MOVECARET_OFFSET = "offset";
    final public static String MOVECARET_POSITION = "position";
    final public static String MOVECARET_STRING = "string";

    public MoveCaretCommand(){
        super(CommandType.MOVECARET);
    }

    @Override
    public void execute(ExecutionList executionList) throws InterruptedException, ExecutionException {
        Element element = executionList.getElements().poll();

        if (element.getAttribute(MOVECARET_OFFSET) != null) {
            final String offsetString = (element.getAttribute(MOVECARET_OFFSET).getValue());
            final int offset = Integer.parseInt(offsetString);

            executionList.getEditor().getCaretModel().moveToOffset(offset);
        } else if (element.getAttribute(MOVECARET_POSITION) != null){
            final String positionString = (element.getAttribute(MOVECARET_POSITION).getValue());
            String[] splitStrings = positionString.split(":");
            assert(splitStrings.length == 2);

            final int line = Integer.parseInt(splitStrings[0]);
            final int column = Integer.parseInt(splitStrings[1]);

            executionList.getEditor().getCaretModel().moveToLogicalPosition(new LogicalPosition(line - 1, column - 1));
        } else if (element.getAttribute(MOVECARET_STRING) != null) {
            final Editor editor = executionList.getEditor();
            final Document document = editor.getDocument();
            final Project project = executionList.getProject();

            final FindManager findManager = FindManager.getInstance(project);
            final FindModel model = findManager.getFindInFileModel().clone();
            model.setGlobal(false);
            model.setReplaceState(false);

            final String value_start = element.getAttribute(MOVECARET_STRING).getValue();
            model.setStringToFind(value_start);
            final FindResult start = FindManager.getInstance(project).findString(document.getCharsSequence(), 0, model);

            executionList.getEditor().getCaretModel().moveToOffset(start.getStartOffset());
        }

        startNextCommand(executionList);
    }
}

File: ide-features-trainer/src/training/commands/MoveCaretCommand.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package training.commands

import com.intellij.find.FindManager
import com.intellij.find.FindResult
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.editor.LogicalPosition
import org.jdom.Element

class MoveCaretCommand : Command(CommandType.MOVECARET) {

  override fun execute(executionList: ExecutionList) {
    val element = executionList.elements.poll()

    ApplicationManager.getApplication().invokeAndWait {
      when {
        element.getAttribute(MOVECARET_OFFSET) != null -> {
          val offsetString = element.getAttribute(MOVECARET_OFFSET)!!.value
          val offset = Integer.parseInt(offsetString)
          executionList.editor.caretModel.moveToOffset(offset)
        }
        element.getAttribute(MOVECARET_POSITION) != null -> {
          val (line, column) = getLineAndColumn(element)
          executionList.editor.caretModel.moveToLogicalPosition(LogicalPosition(line - 1, column - 1))
        }
        element.getAttribute(MOVECARET_STRING) != null -> {
          val start = getStartOffsetForText(executionList, element)
          executionList.editor.caretModel.moveToOffset(start.startOffset)
        }
      }
    }

    startNextCommand(executionList)
  }

  private fun getStartOffsetForText(executionList: ExecutionList, element: Element): FindResult {
    val editor = executionList.editor
    val document = editor.document
    val project = executionList.project

    val findManager = FindManager.getInstance(project)
    val model = findManager.findInFileModel.clone()
    model.isGlobal = false
    model.isReplaceState = false

    val valueStart = element.getAttribute(MOVECARET_STRING)!!.value
    model.stringToFind = valueStart
    return FindManager.getInstance(project).findString(document.charsSequence, 0, model)
  }

  private fun getLineAndColumn(element: Element): Pair<Int, Int> {
    val positionString = element.getAttribute(MOVECARET_POSITION)!!.value
    val splitStrings = positionString.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    assert(splitStrings.size == 2)
    return Pair(Integer.parseInt(splitStrings[0]), Integer.parseInt(splitStrings[1]))
  }

  companion object {

    const val MOVECARET_OFFSET = "offset"
    const val MOVECARET_POSITION = "position"
    const val MOVECARET_STRING = "string"
  }
}
